File: python/src/kora/auto.py
"""KoraAuto — scan-mode SDK for spend observation (no signing, no enforcement).

Emits spend-intent signals to Kora's observation endpoint. Observations are
used by admins to discover candidate agents for delegation.

Errors are logged to stderr (prefix KORA_SCAN_WARN), never raised.
"""
from __future__ import annotations

import os
import socket
import sys
import time
from datetime import datetime, timezone
from typing import Any

import requests


_DEFAULT_BASE_URL = "https://api.koraprotocol.com"


class KoraAuto:
    """Scan-mode SDK — observe spend intent, never sign or enforce.

    Args:
        scan_token: Scan token issued by admin via POST /v1/auto/tokens
        base_url: Kora API base URL
    """

    def __init__(self, scan_token: str, base_url: str = _DEFAULT_BASE_URL):
        self._scan_token = scan_token
        self._base_url = base_url.rstrip("/")
        # Throttle: track (vendor, error_reason) → last_warn_time
        self._warn_throttle: dict[tuple[str, str], float] = {}

    def observe(
        self,
        vendor: str,
        amount_cents: int | None = None,
        currency: str | None = None,
        reason: str | None = None,
        service_name: str | None = None,
        environment: str | None = None,
        runtime_id: str | None = None,
        repo_hint: str | None = None,
    ) -> dict:
        """Emit a spend observation to Kora.

        Auto-detects runtime fields from environment variables if not provided:
        - service_name: KORA_SERVICE_NAME or hostname
        - environment: KORA_ENVIRONMENT or "unknown"
        - runtime_id: KORA_RUNTIME_ID or "unknown"

        Returns {"status": "ok"} on success.
        On failure: logs KORA_SCAN_WARN to stderr, does NOT raise.
        """
        svc = service_name or os.environ.get("KORA_SERVICE_NAME") or socket.gethostname()
        env = environment or os.environ.get("KORA_ENVIRONMENT") or "unknown"
        rt_id = runtime_id or os.environ.get("KORA_RUNTIME_ID") or "unknown"

        body: dict[str, Any] = {
            "signal_type": "EXPLICIT_SPEND_INTENT",
            "observed_at": datetime.now(timezone.utc).isoformat(),
            "runtime": {
                "service_name": svc,
                "environment": env,
                "runtime_id": rt_id,
            },
            "spend": {
                "vendor_id": vendor,
            },
        }

        if repo_hint is not None:
            body["runtime"]["repo_hint"] = repo_hint
        if amount_cents is not None:
            body["spend"]["amount_cents"] = amount_cents
        if currency is not None:
            body["spend"]["currency"] = currency
        if reason is not None:
            body["spend"]["reason"] = reason

        try:
            resp = requests.post(
                f"{self._base_url}/v1/auto/observe",
                json=body,
                headers={"X-Scan-Token": self._scan_token},
                timeout=5,
            )
            if resp.status_code >= 400:
                error_reason = f"http_{resp.status_code}"
                self._warn(vendor, svc, env, rt_id, error_reason)
                return {"status": "error"}
            return resp.json()
        except Exception as exc:
            error_reason = type(exc).__name__.lower()
            self._warn(vendor, svc, env, rt_id, error_reason)
            return {"status": "error"}

    def _warn(self, vendor: str, service: str, env: str,
              runtime_id: str, error: str) -> None:
        """Emit KORA_SCAN_WARN to stderr, throttled to once per 60s per (vendor, error)."""
        key = (vendor, error)
        now = time.monotonic()
        last = self._warn_throttle.get(key, 0.0)
        if now - last < 60.0:
            return
        self._warn_throttle[key] = now

        print(
            f"KORA_SCAN_WARN vendor={vendor} service={service} "
            f"env={env} runtime={runtime_id} error={error}",
            file=sys.stderr,
        )

File: typescript/src/auto.ts
/**
 * KoraAuto — scan-mode SDK for spend observation (no signing, no enforcement).
 *
 * Emits spend-intent signals to Kora's observation endpoint. Observations are
 * used by admins to discover candidate agents for delegation.
 *
 * Errors are logged to stderr (prefix KORA_SCAN_WARN), never thrown.
 */

const DEFAULT_BASE_URL = 'https://api.koraprotocol.com';

export interface KoraAutoConfig {
  scanToken: string;
  baseUrl?: string;
}

export interface ObserveParams {
  vendor: string;
  amountCents?: number;
  currency?: string;
  reason?: string;
  serviceName?: string;
  environment?: string;
  runtimeId?: string;
  repoHint?: string;
}

export class KoraAuto {
  private scanToken: string;
  private baseUrl: string;
  private warnThrottle: Map<string, number> = new Map();

  constructor(config: KoraAutoConfig) {
    this.scanToken = config.scanToken;
    this.baseUrl = (config.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, '');
  }

  async observe(params: ObserveParams): Promise<{ status: string }> {
    const svc =
      params.serviceName ??
      process.env.KORA_SERVICE_NAME ??
      (await import('node:os')).hostname();
    const env =
      params.environment ?? process.env.KORA_ENVIRONMENT ?? 'unknown';
    const rtId =
      params.runtimeId ?? process.env.KORA_RUNTIME_ID ?? 'unknown';

    const body: Record<string, unknown> = {
      signal_type: 'EXPLICIT_SPEND_INTENT',
      observed_at: new Date().toISOString(),
      runtime: {
        service_name: svc,
        environment: env,
        runtime_id: rtId,
        ...(params.repoHint != null ? { repo_hint: params.repoHint } : {}),
      },
      spend: {
        vendor_id: params.vendor,
        ...(params.amountCents != null ? { amount_cents: params.amountCents } : {}),
        ...(params.currency != null ? { currency: params.currency } : {}),
        ...(params.reason != null ? { reason: params.reason } : {}),
      },
    };

    try {
      const resp = await fetch(`${this.baseUrl}/v1/auto/observe`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Scan-Token': this.scanToken,
        },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(5000),
      });

      if (!resp.ok) {
        this.warn(params.vendor, svc, env, rtId, `http_${resp.status}`);
        return { status: 'error' };
      }

      return (await resp.json()) as { status: string };
    } catch (err) {
      const errorReason =
        err instanceof Error ? err.constructor.name.toLowerCase() : 'unknown';
      this.warn(params.vendor, svc, env, rtId, errorReason);
      return { status: 'error' };
    }
  }

  private warn(
    vendor: string,
    service: string,
    env: string,
    runtimeId: string,
    error: string,
  ): void {
    const key = `${vendor}:${error}`;
    const now = Date.now();
    const last = this.warnThrottle.get(key) ?? 0;
    if (now - last < 60_000) return;
    this.warnThrottle.set(key, now);

    process.stderr.write(
      `KORA_SCAN_WARN vendor=${vendor} service=${service} env=${env} runtime=${runtimeId} error=${error}\n`,
    );
  }
}
